### ISSUE TITLE
Scanning of Midi Interface blocks Darktable

### ISSUE THREAD

**Issue by devarni** [darktable-org/darktable#10687]
Scanning the midi devices blocks Darktable on startup... on the midi interface all ports are scanned continuously for ever. If the midi interface is deactivated, Darktable starts normally.
(Tested with ESI M8U XL USB Midi interface)

**Comment by dterrahe** [comment]
Which platform are you running on?
Is anything connected to the interface?
Could you run from the command line with -d input and report any debugging output?
Does adding the line "plugins/midi/devices=-" to darktablerc help? (i.e. allow you to start dt without it hanging while the interface is connected, albeit without any midi inputs being used)

**Comment by devarni** [comment]
Windows 10 and no external  hardware is enabled on the midi interface.
No debug output with the switch "-d input" but adding the line "plugins/midi/devices=-" solves this bug!

It looks like it wouldn't be a big problem to change the test routine a bit and implement an error counter that exits the loop after 2-3 failed attempts?

**Comment by dterrahe** [comment]
Without output from -d input we would not know which loop is causing your issue. Each MIDI device as identified by the portmidi library should only be tried once.

With plugins/midi/devices=- each device is still considered once, but none are opened. So it doesn't seem there is a loop over all devices that never ends. Maybe there is an internal device in the interface that gets opened and sends data continuously. Debug output would really help. Do you run from the command line?

**Comment by devarni** [comment]
C:\Program Files\darktable\bin\darktable.exe -d input
Yes, I run it from the command line.. there is no debug output, is it written to the standard output channel or to a file?

**Comment by devarni** [comment]
Found it... its not written to the standard output but to a file... (attached)
[darktable-log.txt](https://github.com/darktable-org/darktable/files/7776024/darktable-log.txt)

**Comment by dterrahe** [comment]
Many thanks!

It looks like all midi devices get opened only once. Do you have any shortcut mappings set up for any midi inputs? In a complicated setup like this (many in/outputs) there could be a feedback loop. Does a clean shortcutsrc avoid the problem? Or disconnecting just the Steinberg (of just the M8U)?

Separately, i thought the pango problem with msys2 was resolved, but you have many asserts in the log. Did you compile from source? Then maybe updating all libs could help (pacman -Syu). Otherwise the packager should have a look at this.

**Comment by devarni** [comment]
It is the default installation without changing anything, I removed the shortcutsrc which has no effect.
Yes, disconnecting the M8UXL is the only solution (besides adding the extra line in the configuration).
I see on my midi interface that on all ports data is transferred one after the other and in endless loop (looks like a running light on the midi outputs).... 

No, it is the official installer, I did not compile DT myself from source. The Pango error is then another problem with the official Windows package.

**Comment by dterrahe** [comment]
Could also be that the pango issue got introduced with the latest update (I don't have 1.50.2-1 installed yet and don't have the error message). Could also be that it is linked to the fonts you have installed... Anyway, that's a separate issue; and it doesn't seem to cause crashes for you, so may not be worth reporting.

The continuous data you see flowing over midi is the status of all midi notes getting switched off all the time. This allows some controllers to show status lights when their keys are linked to a toggle button; the lights go on when the toggle is switched on. I guess in your case, with 9 separate devices being open, all of those notes take longer than the repeat interval to get sent and lock up darktable.

Since I've received very little feedback about midi devices I do not own myself, I haven't introduced any configurability needed to make those setups work reliably, apart from "plugins/midi/devices". Did you actually want to use midi? I.e. do you have any devices that might/should work with dt? Then you could use plugins/midi/devices= to selectively enable only those devices. Supply a part of the name of the "good" device immediately after the = and then add ;- (semicolon and minus).

**Comment by agittins** [comment]
> The continuous data you see flowing over midi is the status of all midi notes getting switched off all the time.

Does the midi implementation talk to all discovered midi devices by default? This might not be ideal behaviour and doesn't follow principle of least surprise. You never know what people might have hooked up to their midi ports, and you never know what they might do when receiving unexpected data. Unlikely to be a huge issue for synths, but a lighting or fireworks controller might be interesting!

I am super excited to try out the midi support in the next few days though, I have a control surface I'm keen to mess around with :-)

**Comment by dterrahe** [comment]
> This might not be ideal behaviour and doesn't follow principle of least surprise.

You are right of course. I was aiming for an out-of-the-box experience where the majority of devices would Just Work. That would probably require a white list or database of some kind but that did not get implemented due to lack of time and feedback. Midi support should not be considered mature, except for the x-touch mini.
 
Having said that, the only thing that gets sent automatically is repeating note offs. I would hope that the damage this could do is limited and it is the easy way to guarantee that lighted buttons that have been mapped in one view and are switched on do not stay on when switching to another view. 

But I had not considered fireworks controllers. Hopefully their safety protocols prohibit installation of random software that could start sending midi data? Otherwise things could get interesting in the run up to New Year's. Does nuclear equipment use midi?

Control surfaces can be very complicated beasts and may need significant configuration themselves before they start sending (and receiving) data that dt can work with. So I am very interested in your feedback. Both to inform further development, eventually when I'm done with the base input system, and also to add to the docs section to help others get started.

**Comment by TurboGit** [comment]
Fix merged, closing.
